src/services/anchor-service.ts

```diff
@@ -240,188 +256,180 @@ export default class AnchorService {
-      }
+  async _loadCandidateStreams(candidates: Candidate[], candidateLimit: number): Promise<LoadCandidatesResult> {
+    const failedRequests: Request[] = [];
+    const conflictingRequests: Request[] = [];
+    const unprocessedRequests: Request[] = [];
+    const alreadyAnchoredRequests: Request[] = [];
+
+    let index = 0;
+    let numSelectedCandidates = 0;
+    if (candidateLimit == 0) {
+      // 0 means no limit
+      candidateLimit = candidates.length
+    }
 
-      if (stream.tip.toString() != request.cid) {
-        logger.warn(`When loading stream ${stream.id.toString()} at commit ${request.cid}, stream was returned at tip ${stream.tip.toString()}`)
-      }
+    while (index < candidates.length && numSelectedCandidates < candidateLimit) {
+      const batchSize = Math.min(BATCH_SIZE, candidateLimit - numSelectedCandidates)
+      const batchCandidates = candidates.slice(index, Math.min(index + batchSize, candidates.length))
+      index += batchSize;
 
-      return new Candidate(new CID(request.cid), request.id, stream);
-    } catch (e) {
-      logger.err(`Error while loading stream ${streamId?.baseID.toString()} at commit ${streamId?.commit.toString()}. ${e}`)
-      await this.requestRepository.updateRequests({
-        status: RS.FAILED,
-        message: "Request has failed. " + e.toString(),
-      }, [request]);
+      await Promise.all(batchCandidates.map(async (candidate) => {
+        await AnchorService._loadCandidate(candidate, this.ceramicService);
+        if (candidate.shouldAnchor()) {
+          numSelectedCandidates++;
+        }
+        failedRequests.push(...candidate.failedRequests)
+        conflictingRequests.push(...candidate.rejectedRequests)
+        if (candidate.alreadyAnchored) {
+          alreadyAnchoredRequests.push(...candidate.acceptedRequests)
+        }
+      }))
     }
+
+    return { alreadyAnchoredRequests, conflictingRequests, failedRequests, unprocessedRequests }
   }
 
   /**
-   * Selects which Candidate CID should be anchored for each streamId
-   * @param candidates - List of Candidates each representing one anchor request
-   * @return a tuple whose first element is an array of the Candidates that were selected for anchoring,
-   *   and whose second element is an array of Candidates that were rejected by the conflict resolution rules
+   * First starts by loading the CommitID for each pending Request on this Stream. This ensures
+   * that the Ceramic node we are using has at least heard of and considered every commit that
+   * has a pending anchor request. Then we load the current version of the Stream from that Ceramic
+   * node and can then trust that the version of the Stream we receive back is the best, most current
+   * version of the stream based on the node's internal conflict resolution mechanism. We can then
+   * use the current version of the Stream to decide what CID to anchor.
```

Comment:
Can we make this into a multiquery that loads both the *streamid* and the *commitid* in one request? We do this in 3id-did-resolver for example.

Reply:
We discussed this on the original spec document as well.  We can't load the CommitIDs and the StreamID in one multiquery as Ceramic processes all requests in a Multiquery concurrently, so there's no guarantee that the load of the base StreamID will consider all the tips from the CommitIDs.  We could, however, group all the CommitIDs into a single Multiquery and just have the base StreamID load happen after.  I can make that change

Reply:
> so there's no guarantee that the load of the base StreamID will consider all the tips from the CommitIDs

What? That seems like a bug?

Reply:
> What? That seems like a bug?

That's debatable.  It does make some sense for a Multiquery to load all of the component requests concurrently.  We could, however, change MultiQuery to always load all CommitIDs provided first, and then if there are also StreamIDs for the same streams only load them *after* the commitIDs for the same streams have been considered, but that could slow things down for use cases that don't require this behavior of using the CommitID load as a way to inform the node of a tip it may not have heard of.

Reply:
>  We could, however, change MultiQuery to always load all CommitIDs provided first

Why would it need to load CommitIDs first? If you load a CommitID after should we not do conflict resolution if the commit tip is not in the log? 

> but that could slow things down for use cases that don't require this behavior

Could we not have the multiquery system hold handles to the streams that they load and listen for updates until all streams are loaded?

Reply:
> Could we not have the multiquery system hold handles to the streams that they load and listen for updates until all streams are loaded?

Yeah actually something like that could work.  Syncing all loaded streams to the most recent state in the cache right before returning them.  I'll give that a shot on Monday

Reply:
Nice.

Reply:
So thinking about this a bit more, there's a cost to using MultiQueries right now, which I think exposes a bigger issue with the multiQueries api in general.

With the implementation as it's written right now in this PR, it's possible for there to be multiple requests for different tips on the same stream, and for some of those requests to fail while others succeed, and the best tip for the stream still gets anchored.  Now if I switch to using multiqueries, now if loading *any* of the CommmitIDs fails, the whole multiQuery fails, and the only option left is to fail *all* the requests on this stream.

This is an issue for the multiQuery api in general, I'll file a ticket for it.

In the meantime, I think we should merge this as-is.  The CAS isn't very latency sensitive, so a few extra round trips won't be killer, plus we're doing the processing for multiple streams concurrently anyway.

Reply:
oh actually nevermind - I guess multiQuery just swallows any error for any of the individual queries.  That removes this concern then.  I'll go ahead and try to switch over to multiqueries, and I'll file a ticket for more fine-grained error reporting from multiQueries

Reply:
Updated!